### PR TITLE
[MODEL] TransformerXL Language Modeling

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
This adds the TransformerXL model to the language_modeling scripts. As a first step, converting and working with / evaluating pretrained models published by the author is supported. Training will be added as a next step.

Before merging, the pretrained models must be uploaded to S3 and the evaluation script must be extended for 'enwiki8', 'lm1b', 'text8'. Currently only Wikitext103 was converted.

### Essentials ###
- [X] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [X] Pre-trained language modeling with TransformerXL

## Comments ##
This PR includes an efficient implementation of Multi-head attention with relative position embedding as well as adaptive embedding and adaptive softmax. All are HybridBlocks, but the latter two should be improved with control flow ops in the future to avoid unnecessary computation. it thus replaces (?) https://github.com/dmlc/gluon-nlp/pull/671 and https://github.com/dmlc/gluon-nlp/pull/255.

For the first two batches of Wikitext103 testset, all Gluon model outputs match the outputs of the Pytorch implementation of the original author:

```
Batch 0
Layer 0: Maximum error 3.05e-05 at position (1474, 358). stdev=1.14e-06
Layer 1: Maximum error 8.34e-07 at position (1563, 544). stdev=1.73e-08
Layer 2: Maximum error 1.43e-06 at position (1560, 506). stdev=2.38e-08
Layer 3: Maximum error 1.91e-06 at position (1515, 506). stdev=2.54e-08
Layer 4: Maximum error 3.34e-06 at position (1589, 506). stdev=3.66e-08
Layer 5: Maximum error 1.91e-06 at position (1504, 533). stdev=3.93e-08
Layer 6: Maximum error 1.91e-06 at position (1504, 533). stdev=4.96e-08
Layer 7: Maximum error 1.91e-06 at position (1529, 506). stdev=4.29e-08
Layer 8: Maximum error 2.38e-06 at position (1504, 533). stdev=4.55e-08
Layer 9: Maximum error 1.91e-06 at position (1568, 506). stdev=4.63e-08
Layer 10: Maximum error 1.91e-06 at position (1504, 533). stdev=5.81e-08
Layer 11: Maximum error 1.91e-06 at position (1568, 506). stdev=5.91e-08
Layer 12: Maximum error 2.52e-06 at position (1476, 350). stdev=7.42e-08
Layer 13: Maximum error 2.32e-06 at position (1544, 734). stdev=1.09e-07
Layer 14: Maximum error 3.49e-06 at position (1504, 553). stdev=1.51e-07
Layer 15: Maximum error 3.58e-06 at position (1597, 120). stdev=1.46e-07
Layer 16: Maximum error 3.46e-06 at position (1533, 543). stdev=1.47e-07
Layer 17: Maximum error 3.34e-06 at position (1522, 137). stdev=1.58e-07
Loss: Maximum error 1.91e-06 at position (70,). stdev=5.68e-06
Batch 1
Layer 0: Maximum error 3.43e-05 at position (1599, 328). stdev=1.64e-06
Layer 1: Maximum error 9.54e-07 at position (1502, 147). stdev=2.65e-08
Layer 2: Maximum error 1.43e-06 at position (1432, 506). stdev=3.47e-08
Layer 3: Maximum error 1.91e-06 at position (1387, 506). stdev=3.57e-08
Layer 4: Maximum error 3.81e-06 at position (1477, 506). stdev=5.10e-08
Layer 5: Maximum error 2.38e-06 at position (1484, 369). stdev=5.58e-08
Layer 6: Maximum error 2.38e-06 at position (1484, 369). stdev=6.85e-08
Layer 7: Maximum error 1.91e-06 at position (1401, 506). stdev=5.93e-08
Layer 8: Maximum error 2.38e-06 at position (1376, 533). stdev=6.37e-08
Layer 9: Maximum error 1.91e-06 at position (1440, 506). stdev=6.52e-08
Layer 10: Maximum error 1.91e-06 at position (1376, 533). stdev=8.15e-08
Layer 11: Maximum error 2.29e-06 at position (1494, 406). stdev=8.40e-08
Layer 12: Maximum error 2.91e-06 at position (1494, 406). stdev=1.04e-07
Layer 13: Maximum error 3.13e-06 at position (1494, 406). stdev=1.52e-07
Layer 14: Maximum error 3.76e-06 at position (1473, 872). stdev=2.22e-07
Layer 15: Maximum error 3.58e-06 at position (1469, 120). stdev=2.09e-07
Layer 16: Maximum error 3.46e-06 at position (1405, 543). stdev=2.11e-07
Layer 17: Maximum error 5.25e-06 at position (1580, 511). stdev=2.27e-07
Loss: Maximum error 3.34e-06 at position (54,). stdev=8.68e-06
```

### CODE REVIEW

**Comment by leezu** [comment]
Rebasing on master to include CI fixes

**Comment by mli** [comment]
Found link check problems in job PR-846/3:
(line   19) broken    https://www.salesforce.com/products/einstein/ai-research/the-wikitext-dependency-language-modeling-dataset/ - 404 Client Error: Not Found for url: https://www.salesforce.com/products/einstein/ai-research/the-wikitext-dependency-language-modeling-dataset/
(line    5) broken    https://www.salesforce.com/products/einstein/ai-research/the-wikitext-dependency-language-modeling-dataset/ - 404 Client Error: Not Found for url: https://www.salesforce.com/products/einstein/ai-research/the-wikitext-dependency-language-modeling-dataset/
(line    5) broken    https://www.salesforce.com/products/einstein/ai-research/the-wikitext-dependency-language-modeling-dataset/ - 404 Client Error: Not Found for url: https://www.salesforce.com/products/einstein/ai-research/the-wikitext-dependency-language-modeling-dataset/
(line   21) broken    https://www.salesforce.com/products/einstein/ai-research/the-wikitext-dependency-language-modeling-dataset/ - 404 Client Error: Not Found for url: https://www.salesforce.com/products/einstein/ai-research/the-wikitext-dependency-language-modeling-dataset/
(line  442) broken    https://www.aclweb.org/anthology/P02-1040.pdf)[1 - 404 Client Error: Not Found for url: https://www.aclweb.org/anthology/P02-1040.pdf)%5B1
(line   92) broken    https://nlp.stanford.edu/pubs/glove.pdf)[2 - 404 Client Error: NOT FOUND for url: https://nlp.stanford.edu/pubs/glove.pdf)%5B2
(line  208) broken    https://www.bioinf.jku.at/publications/older/2604.pdf)[3 - 404 Client Error: Not Found for url: https://www.bioinf.jku.at/publications/older/2604.pdf)%5B3

**Comment by mli** [comment]
Job PR-846/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-846/3/index.html

File: scripts/language_model/transformer/model.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+from .transformer import TransformerXL
+
+__all__ = ['get_model']
+
+
+def get_model(name, **kwargs):
+    """Returns a pre-defined model by name."""
+    models = {
+        # TODO better naming scheme when moving this to main API?
+        'transformerxl': transformerxl,
```
Comment (eric-haibin-lin) [inline scripts/language_model/transformer/model.py]: maybe transformer_xl?

File: scripts/language_model/transformer/softmax.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+class AdaptiveLogSoftmaxWithLoss(mx.gluon.HybridBlock):
+    """Efficient softmax approximation
+
+    Grave, E., Joulin, A., Cissé, M., Jégou, H., & others, (2017). Efficient
+    softmax approximation for GPUs. In , Proceedings of the 34th International
+    Conference on Machine Learning-Volume 70 (pp. 1302–1310).
+
+    Parameters
+    ----------
+    vocab_size
```
Comment (eric-haibin-lin) [inline scripts/language_model/transformer/softmax.py]: document type, too
vocab_size : int

File: scripts/language_model/transformer/model.py (hunk 1 above)
Comment (leezu) [reply scripts/language_model/transformer/model.py]: The problem here is that different architectures are used under the same name "transformer_xl" for different datasets. We could postfix the dataset to the name (eg. `transformer_xl_wt103`, `transformer_xl_text8` etc.), but this would make the `dataset=` argument of our `get_model` function quite awkward.
Naming all the different models the same and returning different models depending on the dataset name is possible but I think we haven't done this so far. It may be unexpected.

File: scripts/language_model/transformer/softmax.py (hunk 2 above)
Comment (leezu) [reply scripts/language_model/transformer/softmax.py]: Type will be documented automatically (https://github.com/dmlc/gluon-nlp/pull/830)

File: scripts/language_model/transformer/model.py (hunk 1 above)
Comment (eric-haibin-lin) [reply scripts/language_model/transformer/model.py]: Are these architectures different due to different embedding dimensions, or something else?

Comment (leezu) [reply scripts/language_model/transformer/model.py]: They vary across several parameters, such as number of layers, dimensions, which parameters are shared, how adaptive embedding / softmax is set up, ...: https://github.com/kimiyoung/transformer-xl/blob/master/tf/sota/enwik8.sh https://github.com/kimiyoung/transformer-xl/blob/master/tf/sota/lm1b.sh https://github.com/kimiyoung/transformer-xl/blob/master/tf/sota/text8.sh https://github.com/kimiyoung/transformer-xl/blob/master/tf/sota/wt103.sh

**Comment by mli** [comment]
Job PR-846/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-846/5/index.html

**Comment by mli** [comment]
Job PR-846/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-846/6/index.html

**Comment by mli** [comment]
Job PR-846/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-846/7/index.html
